src/Adapter/Product/ProductDuplicator.php

```diff
@@ -0,0 +1,637 @@
+    {
+        //@todo: add database transaction. After/if PR #21740 gets merged
+        $product = $this->productRepository->get($productId);
+        $oldProductId = $productId->getValue();
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateBefore',
+            ['id_product' => $oldProductId]
+        );
+        $newProduct = $this->duplicateProduct($product);
+        $newProductId = (int) $newProduct->id;
+
+        $this->duplicateRelations($oldProductId, $newProductId);
+
+        if ($product->hasAttributes()) {
+            $this->updateDefaultAttribute($newProductId, $oldProductId);
+        }
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionProductAdd',
+            ['id_product_old' => $oldProductId, 'id_product' => $newProductId, 'product' => $newProduct]
+        );
+
+        $this->updateSearchIndexation($newProduct, $oldProductId);
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateAfter',
+            ['id_product' => $oldProductId, 'id_product_new' => $newProductId]
+        );
+
+        return new ProductId((int) $newProduct->id);
+    }
+
+    /**
+     * @param Product $newProduct
+     * @param int $oldProductId
+     *
+     * @throws CannotUpdateProductException
+     */
+    private function updateSearchIndexation(Product $newProduct, int $oldProductId): void
+    {
+        $productIsVisibleInSearch = in_array(
+            $newProduct->visibility,
+            [ProductVisibility::VISIBLE_EVERYWHERE, ProductVisibility::VISIBLE_IN_SEARCH]
+        );
+
+        if (!$this->isSearchIndexationOn || !$productIsVisibleInSearch) {
+            return;
+        }
+
+        try {
+            if (!Search::indexation(false, $newProduct->id)) {
+                throw new CannotUpdateProductException(
+                    sprintf('Cannot update search indexation when duplicating product %d', $oldProductId),
+                    CannotUpdateProductException::FAILED_UPDATE_SEARCH_INDEXATION
+                );
+            }
+        } catch (PrestaShopException $e) {
+            throw new CoreException(
+                sprintf('Error occurred when duplicating product %d. Failed to update search indexation', $oldProductId),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return Product the new product
+     */
+    private function duplicateProduct(Product $product): Product
+    {
+        $this->setName($product);
+        $this->setPriceByShops($product);
+
+        $product->indexed = false;
+        $product->active = false;
+
+        return $this->productRepository->duplicate($product);
+    }
+
+    /**
+     * Sets new name for duplicated product by modifying the existing name with a new pattern
+     *
+     * @param Product $product
+     */
+    private function setName(Product $product): void
```

Comment:
I have been a bit confused when reading this function 😮  because most of the time mutator functions like `setX()` do modify the instance, not the parameter.

May I suggest
```suggestion
    private function modifyProductName(Product $product): void
```
or
```suggestion
    private function modifyNameForDuplicate(Product $product): void
```

Reply:
modified the method to build the name and return it instead of setting it to object inside.
Now it makes more sense